src/consensus/tower.zig

```diff
@@ -1982,136 +1470,729 @@ pub const Tower = struct {
+// might pop off a lot of the votes in the tower. The stake from these votes
+// would have rolled up to earlier votes in the tower, which presumably
+// could have helped us pass the threshold check. Worst case, we'll just
+// recheck later without having increased lockouts.
+fn optimisticallyBypassVoteStakeThresholdCheck(
+    // Needs to be an iterator that produces Lockout
+    tower_before_applying_vote: anytype,
+    threshold_vote: Lockout,
+) bool {
+    for (tower_before_applying_vote.constSlice()) |old_vote| {
+        if (old_vote.slot == threshold_vote.slot and
+            old_vote.confirmation_count == threshold_vote.confirmation_count)
+        {
+            return true;
+        }
+    }
+    return false;
+}
 
-        const reset_slot: ?Slot = heaviest_slot;
-        return CandidateVoteAndResetSlots{
-            .candidate_vote_slot = null,
-            .reset_slot = reset_slot,
-            .switch_fork_decision = initial_switch_fork_decision,
+fn checkVoteStakeThreshold(
+    logger: sig.trace.Logger,
+    maybe_threshold_vote: ?Lockout,
+    tower_before_applying_vote: anytype,
+    threshold_depth: usize,
+    threshold_size: f64,
+    slot: Slot,
+    voted_stakes: *const AutoHashMapUnmanaged(Slot, u64),
+    total_stake: u64,
+) ThresholdDecision {
+    const threshold_vote = maybe_threshold_vote orelse {
+        // Tower isn't that deep.
+        return ThresholdDecision{ .passed_threshold = {} };
+    };
+
+    const fork_stake = voted_stakes.get(threshold_vote.slot) orelse {
+        // We haven't seen any votes on this fork yet, so no stake
+        return ThresholdDecision{
+            .failed_threshold = .{ threshold_depth, 0 },
         };
+    };
+
+    const lockout = @as(f64, @floatFromInt(fork_stake)) / @as(
+        f64,
+        @floatFromInt(total_stake),
+    );
+
+    logger.trace().logf(
+        \\fork_stake slot: {}, threshold_vote slot: {}, lockout: {} fork_stake:
+        \\{} total_stake: {}
+    ,
+        .{
+            slot,
+            threshold_vote.slot,
+            lockout,
+            fork_stake,
+            total_stake,
+        },
+    );
+
+    if (optimisticallyBypassVoteStakeThresholdCheck(
+        tower_before_applying_vote,
+        threshold_vote,
+    ) or lockout > threshold_size) {
+        return ThresholdDecision{ .passed_threshold = {} };
     }
-};
+
+    return ThresholdDecision{
+        .failed_threshold = .{ threshold_depth, 0 },
+    };
+}
+
+pub fn populateAncestorVotedStakes(
+    voted_stakes: *SortedSet(Slot),
+    vote_slots: []const Slot,
+    ancestors: *const AutoHashMapUnmanaged(Slot, SortedSet(Slot)),
+) !void {
+    // If there's no ancestors, that means this slot must be from before the current root,
+    // in which case the lockouts won't be calculated in bank_weight anyways, so ignore
+    // this slot
+    for (vote_slots) |vote_slot| {
+        if (ancestors.get(vote_slot)) |maybe_slot_ancestors| {
```

Comment:
Should perhaps be using `ancestors.getPtr` so that the copy isn't necessary, and so that the sort set is actually be sorted?

Reply:
[87a06af](https://github.com/Syndica/sig/pull/724/commits/87a06af898e3faa6cb3745d438fe4be1d43350cd)